# Student Perfommance Indicator

#### Life Cycle of Machine Learning Project


* Understanding the Problem Statement
* Data Collection 
* Data Checks to perfrom
* Exploratory Data Analysis
* Data Pre-Processing
* Model Training
* Choose Beset model


### 1. Probelm Statement

* This project understands how the student's performance (test scores) is affected by the other varibales such as Gender, Ethinicity, Parental level of education, Lunch and Test preparation course/

### 2. Data Collection 

* Dataset Source - student.csv
* The data consists of 8 columns and 1000 rows

### 2.1 Import Data and Required Packages

#### importing Pandas, Numpy, Matplotib, Seaborn and Warnings Library.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

#### Importing the student.csv file as Pandas DataFrame

In [2]:
df = pd.read_csv('data/student.csv')

In [3]:
df.shape

(1000, 8)

### 2.2 Dataset Information

* gender: sex of students->(Male/Female)
* race/ethnicity: ethnicity of students -> (Group A, B, C, D, E)
* Parental level of Education : parent final education -> ( Bachelor's degree, some college, master's degree, associate's degree, high school)
* Lunch: having lunch before test (standard or free/ reduced)
* test preparation course: complete or not complete before test
* math score
* reading score
* writing score

### 3. Data Checks to perform

* Check Missing Values
* Check Duplicates
* Check data type
* Check the number of unique values of each column
* Check statistics of data set
* Check various categories present in the different categorical columns

In [4]:
# Check for Missing Values
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

There are no missing values in the data set

In [5]:
## Check for Duplicates
df.duplicated().sum()

np.int64(0)

In [6]:
# Check data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [7]:
## Checking the number of unique values of each column
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

In [8]:
## Checking the statistics of the dataset
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

#### Insight

* From above description of numerical data, all means are very close to each other between 66 and 68.05
* All standard deviations are also close between 14.6 and 15.19
* While there is a minimum score 0 for math, for wirting minimum is much higher = 10 and for reading might higher = 17

In [10]:
## Exploring data 
print ("Categories in 'gender' variable:  ", end = " ")
print(df['gender'].unique())

print ("Categories in 'race_ethnicity' variable:  ", end = " ")
print(df['race_ethnicity'].unique())

print ("Categories in 'parental_level_of_education' variable:  ", end = " ")
print(df['parental_level_of_education'].unique())

print ("Categories in 'lunch' variable:  ", end = " ")
print(df['lunch'].unique())

print ("Categories in 'test_preparation_course' variable:  ", end = " ")
print(df['test_preparation_course'].unique())

Categories in 'gender' variable:   ['female' 'male']
Categories in 'race_ethnicity' variable:   ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in 'parental_level_of_education' variable:   ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in 'lunch' variable:   ['standard' 'free/reduced']
Categories in 'test_preparation_course' variable:   ['none' 'completed']


In [11]:
# Define numerical and categorical features
numeric_features = [feature for feature in df.columns if df[feature].dtypes != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtypes == 'O']

print('There are {} numerical features: {}'.format(len(numeric_features), numeric_features))
print('There are {} categorical features: {}'.format(len(categorical_features), categorical_features))

There are 3 numerical features: ['math_score', 'reading_score', 'writing_score']
There are 5 categorical features: ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


In [12]:
## Adding columns for 'Total_Score', 'Average_Score' 
df['Total_Score'] = df['math_score'] + df['reading_score'] + df['writing_score']
df['Average_Score'] = df['Total_Score'] / 3
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   Total_Score  Average_Score  
0          218      72.666667  
1          247      82.333333  
2          278      92.666667  
3          148      49.333333  
4          229      76.333333

In [17]:
reading_full = df[df['reading_score'] == 100]['Average_Score'].count()
writing_full = df[df['writing_score'] == 100]['Average_Score'].count()
math_full = df[df['math_score'] == 100]['Average_Score'].count()
print(reading_full, writing_full, math_full)

17 14 7
